### PR TITLE
Refactor: Move login out of models

### PR DESCRIPTION
**Pull request by zeripath**
`models` does far too much. In particular it handles all `UserSignin`.

It shouldn't be responsible for calling LDAP, SMTP or PAM for signing in.

Therefore we should move this code out of `models`.

This code has to depend on `models` - therefore it belongs in `services`.

There is a package in `services` called `auth` and clearly this functionality belongs in there.

Plan:

- [x] Change `auth.Auth` to `auth.Method` - as they represent methods of authentication.
- [x] Move `models.UserSignIn` into `auth`
- [x] Move `models.ExternalUserLogin`
- [x] Move most of the `LoginVia*` methods to `auth` or subpackages
- [x] Move Resynchronize functionality to `auth`
  - Involved some restructuring of `models/ssh_key.go` to reduce the size of this massive file and simplify its files.
- [x] Move the rest of the LDAP functionality in to the ldap subpackage
- [x] Re-factor the login sources to express an interfaces `auth.Source`?
  - I've done this through some smaller interfaces Authenticator and Synchronizable - which would allow us to extend things in future
- [x] Now LDAP is out of models - need to think about modules/auth/ldap and I think all of that functionality might just be moveable
- [x] Similarly a lot Oauth2 functionality need not be in models too and should be moved to services/auth/source/oauth2
  - [x] modules/auth/oauth2/oauth2.go uses xorm... This is naughty - probably need to move this into models.
  - [x] models/oauth2.go - mostly should be in modules/auth/oauth2 or services/auth/source/oauth2 
- [x] More simplifications of login_source.go may need to be done
- Allow wiring in of notify registration -  *this can now easily be done - but I think we should do it in another PR*  - see #16178 
- More refactors...?
  - OpenID should probably become an auth Method but I think that can be left for another PR
  - Methods should also probably be cleaned up  - again another PR I think.
  - SSPI still needs more refactors.

### CODE REVIEW

**Comment by zeripath** [comment]
OK I think that's enough for now.

Questions:

* Do we think the source subdirectory is required or should things be moved back done to subdirectories of auth directly?
* `services/auth/source/ldap.Source` has a weird wrapper in-order to keep backward compatibility. This could be dropped at the cost of a migration.
* `models.LoginSource.IsActived` should be renamed to `IsActive` but this would require a migration.

File: models/ssh_key_authorized_keys.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
```
Comment (lunny) [inline models/ssh_key_authorized_keys.go:1]: SSH related files should also be moved out of models

Comment (zeripath) [reply models/ssh_key_authorized_keys.go:1]: One thing at a time...

Comment (lunny) [reply models/ssh_key_authorized_keys.go:1]: But you have split them from old file?

Comment (zeripath) [reply models/ssh_key_authorized_keys.go:1]: On the advice from @6543 and to prepare for that move.

I think this is already going to be difficult enough to get reviewed without adding in another refactor but if you'd like it in here I can do that.

The issue is that it'll need to be one of the first things merged in 1.16 if not the first thing as it'll be too easy conflict.

Comment (lunny) [reply models/ssh_key_authorized_keys.go:1]: I think we can keep ssh related code unchanged in this PR or move them to another PR.

Comment (zeripath) [reply models/ssh_key_authorized_keys.go:1]: The changes that have been made are necessary for this PR.

**Comment by lafriks** [comment]
> * `services/auth/source/ldap.Source` has a weird wrapper in-order to keep backward compatibility. This could be dropped at the cost of a migration.

I would prefer migration to keep code more clean

**Comment by zeripath** [comment]
migration added.

**Comment by zeripath** [comment]
updated to include #16465

**Review by 6543 — APPROVED** [review]

**Review by khmarbaise — APPROVED** [review]

**Comment by zeripath** [comment]
make lgtm work

**Pull request merged**